### PR TITLE
[test] Size Performance Regression Tests (SPRT)

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #948

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

The suggested workflow to do regression test for a PR in **reviewer-major** manner is:
* When a performance related PR is ready, checkout that PR locally.
* Run ``ti benchmark && ti regression`` to obtain the result.
* Decide wheater to approve or request change, depends on the result.
* Right before merge, run ``ti baseline`` to save the benchmark result as new baseline.

The suggested workflow to do regression test for a PR in **author-major** manner is:
* When a stage of work is done, run ``ti benchmark && ti regression`` to obtain result.
* When ready for review, post the latest result in PR comments, wait for the reviewer.
* Get feedback from the reviewer, and improve the result.
* Right before merge, run ``ti baseline`` to save the benchmark result as new baseline.

### CODE REVIEW

File: taichi/transforms/constant_fold.cpp [hunk 1]
```diff
@@ -198,6 +198,8 @@ class ConstantFold : public BasicStmtVisitor {
 namespace irpass {
 
 void constant_fold(IRNode *root) {
+  return; // XXX: A H A!
```
Comment (k-ye) [inline taichi/transforms/constant_fold.cpp]: Remove?

File: baselines/test_ad_basics__test_ad_precision_1.dat [hunk 2]
```diff
@@ -0,0 +1,4 @@
+codegen_accessor_statements: 36.00
```
Comment (k-ye) [inline baselines/test_ad_basics__test_ad_precision_1.dat]: Are these the number of statements? Can we print them as integers?

**Comment by archibate** [comment]
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-n80gwu55
[Taichi] sandbox prepared
[Taichi] <dev mode>, supported archs: [cpu, cuda, opengl], commit 2713731e, python 3.8.2

 *******************************************
 **     Taichi Programming Language       **
 *******************************************

Running on Arch=x64

Running python tests...

bringing up nodes...
.                                                                                     [100%]
1 passed in 0.30s
Starting 4 testing thread(s)...
==================================== test session starts ====================================
platform linux -- Python 3.8.2, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /home/bate/Develop/taichi
plugins: rerunfailures-9.0, xdist-1.31.0, forked-1.1.3
gw0 [6] / gw1 [6] / gw2 [6] / gw3 [6]
......                                                                                [100%]
==================================== 6 passed in 17.42s =====================================
linalg__________________polar_decomp______________________________
codegen_statements_________________________________3162.0_3179.0_+
codegen_kernel_statements__________________________2788.0_2819.0_+
codegen_offloaded_tasks______________________________39.0___37.0_-
codegen_accessor_statements_________________________374.0__360.0_-

linalg__________________init_matrix_from_vectors__________________
codegen_statements_________________________________3163.0_3180.0_+
codegen_kernel_statements__________________________2789.0_2820.0_+
codegen_offloaded_tasks______________________________39.0___37.0_-
codegen_accessor_statements_________________________374.0__360.0_-


>>> Running time: 17.99s
```

File: baselines/test_ad_basics__test_ad_precision_1.dat (hunk 2 above)
Comment (archibate) [reply baselines/test_ad_basics__test_ad_precision_1.dat]: Thank for pointing out, yes, we can print then as int, but store them in float was not my decision.

Comment (k-ye) [reply baselines/test_ad_basics__test_ad_precision_1.dat]: i guess it's easier for calculating percentages using floats? @xumingkuan

**Comment by archibate** [comment]
Thank for the suggestions! Looks better now:
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-gmai476d
[Taichi] sandbox prepared
[Taichi] <dev mode>, supported archs: [cpu, cuda, opengl], commit 2713731e, python 3.8.2

 *******************************************
 **     Taichi Programming Language       **
 *******************************************

linalg__________________polar_decomp______________________________
codegen_accessor_statements______________________360___374___-3.7%
codegen_offloaded_tasks___________________________37____39___-5.1%
codegen_statements______________________________3179__3162___+0.5%
codegen_kernel_statements_______________________2819__2788___+1.1%

linalg__________________init_matrix_from_vectors__________________
codegen_accessor_statements______________________360___374___-3.7%
codegen_offloaded_tasks___________________________37____39___-5.1%
codegen_statements______________________________3180__3163___+0.5%
codegen_kernel_statements_______________________2820__2789___+1.1%
```

File: python/taichi/main.py [hunk 3]
```diff
@@ -69,6 +70,58 @@ def test_cpp(args):
     return int(task.run(*test_files))
 
 
+def get_benchmark_baseline_dir():
+    import taichi as ti
+    return os.path.join(ti.core.get_repo_dir(), 'baselines')
```
Comment (k-ye) [inline python/taichi/main.py:75]: BTW, can we place these data not in the root directories? (I don't think these data deserve a slot in the root dir..)

E.g. maybe `benchmark/data/baseline`? Also, I'd suggest using the singular form, i.e. `baseline` and `output`

File: baselines/test_ad_basics__test_ad_precision_1.dat (hunk 2 above)
Comment (xumingkuan) [reply baselines/test_ad_basics__test_ad_precision_1.dat]: Floats are just used for all statistics here.

https://github.com/taichi-dev/taichi/blob/ab2f5dbbf6241b8870ec140a32d4a4e98e78c407/taichi/util/statistics.h#L8

**Comment by xumingkuan** [comment]
> Thank for the suggestions! Looks better now:
> 
> ```
> [Taichi] mode=development
> [Taichi] preparing sandbox at /tmp/taichi-gmai476d
> [Taichi] sandbox prepared
> [Taichi] <dev mode>, supported archs: [cpu, cuda, opengl], commit 2713731e, python 3.8.2
> 
>  *******************************************
>  **     Taichi Programming Language       **
>  *******************************************
> 
> linalg__________________polar_decomp______________________________
> codegen_accessor_statements______________________360___374___-3.7%
> codegen_offloaded_tasks___________________________37____39___-5.1%
> codegen_statements______________________________3179__3162___+0.5%
> codegen_kernel_statements_______________________2819__2788___+1.1%
> 
> linalg__________________init_matrix_from_vectors__________________
> codegen_accessor_statements______________________360___374___-3.7%
> codegen_offloaded_tasks___________________________37____39___-5.1%
> codegen_statements______________________________3180__3163___+0.5%
> codegen_kernel_statements_______________________2820__2789___+1.1%
> ```

Well... I don't understand if the number on the left or the right is the baseline, and if `-` or `+` denotes improvement. `360 -> 374` looks better to me if the left is the baseline.

I think for the numbers which are the smaller the better, `-` denoting improvement looks more reasonable to me.

File: python/taichi/main.py (hunk 3 above)
Comment (xumingkuan) [reply python/taichi/main.py:75]: As we already have the `benchmarks` folder, I agree to put the files in the folder.

We have many folders with plural form names. Is using singular form a better coding style?

Comment (k-ye) [reply python/taichi/main.py:75]: > We have many folders with plural form names. Is using singular form a better coding style?

Feels a bit weird to see `outputs`, but this isn't a big deal.. I think the important part is to not put too much under the root directory

**Comment by archibate** [comment]
> if the left is the baseline.

Yes :)
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-6fo80xax
[Taichi] sandbox prepared
[Taichi] <dev mode>, supported archs: [cpu, cuda, opengl], commit 80319b4d, python 3.8.2

 *******************************************
 **     Taichi Programming Language       **
 *******************************************

linalg__________________polar_decomp______________________________
codegen_kernel_statements                   2819 ->  2788    +1.1%
codegen_accessor_statements                  360 ->   374    -3.7%
codegen_offloaded_tasks                       37 ->    39    -5.1%
codegen_statements                          3179 ->  3162    +0.5%

linalg__________________init_matrix_from_vectors__________________
codegen_kernel_statements                   2820 ->  2789    +1.1%
codegen_accessor_statements                  360 ->   374    -3.7%
codegen_offloaded_tasks                       37 ->    39    -5.1%
codegen_statements                          3180 ->  3163    +0.5%
```
> `-` denoting improvement looks more reasonable to me.

Note although can't see from markdown, these actually have color in terminal: `-` are green for improvement, and '+' are red for regression.

**Comment by xumingkuan** [comment]
> Note although can't be seen from markdown, these actually have colors in terminal: `-` are green for improvement, and `+` are red for regression.

Cool!

> codegen_kernel_statements                   2819 ->  2788    +1.1%

I don't understand: the number decreases from the baseline, isn't it an improvement?

Or if the case is, the baseline is the most recent one, and we're testing previous versions?

**Comment by archibate** [comment]
> > codegen_kernel_statements                   2819 ->  2788    +1.1%
> 
> I don't understand: the number decreases from the baseline, isn't it an improvement?

Sorry! I made an typo, fixed now.

Also note that we have regressions after turns on CFD because, it launches several JIT kernels, which is also accounted in `codegen_kernel_statements`. Should we forbid `codegen_kernel_statements` from recording JIT kernels?

**Comment by xumingkuan** [comment]
> Btw, in fact we have some regressions after turns on CFD because, it launches several JIT kernels, which is also accounted in `codegen_kernel_statements`. Should we forbid `codegen_kernel_statements` from recording JIT kernels?

Sounds good.
BTW, I think they are accessory kernels but not accessor kernels (which help access snodes).

**Comment by archibate** [comment]
Ok, I added `is_evaulator` to `Kernel`:
```
linalg__________________polar_decomp______________________________
codegen_offloaded_tasks                       37 ->    39    +5.4%
codegen_statements                          3179 ->  3162    -0.5%
codegen_kernel_statements                   2819 ->  2788    -1.1%
codegen_evaluator_statements                   0 ->    14    +inf%

linalg__________________init_matrix_from_vectors__________________
codegen_offloaded_tasks                       37 ->    39    +5.4%
codegen_statements                          3180 ->  3163    -0.5%
codegen_kernel_statements                   2820 ->  2789    -1.1%
codegen_evaluator_statements                   0 ->    14    +inf%

```

**Comment by archibate** [comment]
Btw, how do you make the graph/chart? Is that in `misc/`? I guess we can migrate to `ti regression_gui` now.

**Comment by xumingkuan** [comment]
> Btw, how do you make the graph/chart? Is that in `misc/`? I guess we can migrate to `ti regression_gui` now.

I'm just using `matplotlib.pyplot` locally. You can draw one by yourself -- I don't think my code about `advanced_optimization` is kind of reusable. The core code is
```python
plt.figure(figsize=(9, 6))
plt.rcParams.update({'font.size': 16})
plt.bar(range(len(sort_opt)), sort_opt)
plt.ylabel('Remaining statements')
plt.xlabel('{} tests ({} unchanged)'.format(n, cnt_1))
plt.title('Remaining statements after global optimization: {:.3%}'
          .format(opt_mul))
plt.show()
```

**Comment by archibate** [comment]
> LGTM, thanks! Could you also update the PR description of the complete workflow from benchmarking to updating the baseline data to showing the result?

I guess instead of putting in the PR desc, it's better to put it in the docs?

**Comment by k-ye** [comment]
> it's better to put it in the docs?

Yeah, or both...

**Comment by archibate** [comment]
Added, two manner to do PRT: author-major and reviewer-major.

File: baselines/test_ad_basics__test_ad_precision_1.dat (hunk 2 above)
Comment (yuanming-hu) [reply baselines/test_ad_basics__test_ad_precision_1.dat]: Considering that we may need to support float-point numbers in the `Statistics` class, I simply used `float64` when designing this class...

File: docs/utilities.rst [hunk 4]
```diff
@@ -18,8 +18,64 @@ Logging
...
+    codegen_evaluator_statements                   0 ->    14    +inf%
+
+.. note::
+
+    Currently ``ti benchmark`` only support benchmarking number-of-statements, no time benchmarking is included since it depends on hardware performance and therefore hard to compare if the baseline is from another machine.
+    We are to purchase a fixed-performance machine as a time benchmark server at some point.
+    Discussion at: https://github.com/taichi-dev/taichi/issue/948
+
+
+The suggested workflow to do regression test for a PR in **reviewer-major** manner is:
```
Comment (k-ye) [inline docs/utilities.rst:58]: My feeling is that, for most of the times, the reviewers can simply ask the author to run and update the test baselines. So how about this:

```
For PRs that are deemed performance sensitive (e.g. an IR optimization), it's recommended that **the PR author** to run the regression tests, update and upload the baseline results.

The suggested workflow for **the PR author** to run the regression tests is:
...

**Reviewers** can also ask the PR author to run and upload the test result, if they consider the PR performance sensitive. If the reviewers would like to run the regression tests themselves, the suggested workflow is:

...
```

**Comment by archibate** [comment]
> Thanks! This is a very useful feature to have. Just one concern: I compressed all the `dat` files under `benchmarks/baseline` and the package is `115.9KB`, which is also roughly the binary size that goes into `git`. I'm afraid these files will make this repository growing big very quickly. I know using git submodules will make this less automatic, but should we consider creating something like `taichi-dev/taichi_SPRT` to host these day files?

Good idea. I would like to name it `taichi-dev/taichi-baseline` - I use `-` because its like `taichi-docs-zh-cn` or `taichi_elements-docs-zh-cn`. It belongs to `taichi` or `taichi_elements`, instead of a standalone repo, wdyt?

**Comment by yuanming-hu** [comment]
> > Thanks! This is a very useful feature to have. Just one concern: I compressed all the `dat` files under `benchmarks/baseline` and the package is `115.9KB`, which is also roughly the binary size that goes into `git`. I'm afraid these files will make this repository growing big very quickly. I know using git submodules will make this less automatic, but should we consider creating something like `taichi-dev/taichi_SPRT` to host these day files?
> 
> Good idea. I would like to name it `taichi-dev/taichi-baseline` - I use `-` because its like `taichi-docs-zh-cn` or `taichi_elements-docs-zh-cn`. It belongs to `taichi` or `taichi_elements`, instead of a standalone repo, wdyt?

Great idea!! Thanks.

**Comment by archibate** [comment]
> Just one concern: I compressed all the dat files under benchmarks/baseline and the package is 115.9KB,

TBH I don't really think it's a big deal, since `docs` is also very big, and we didn't make it `taichi-docs`. Also note that the size of `baseline` depends only on `tests`.

**Comment by archibate** [comment]
Note that with this attempt, the PR author can't update `baseline` themselves since they usually don't have access to `taichi-baseline`.

**Comment by archibate** [comment]
I added write access for @yuanming-hu @xumingkuan @k-ye, note this means PRs by anyone out of this list will have to use the **reviewer-major** manner as described in the PR desc.
Again, I dislike the submodule attempt, although it helps disable conflict when benchmark dats get merged.

**Comment by yuanming-hu** [comment]
> > Just one concern: I compressed all the dat files under benchmarks/baseline and the package is 115.9KB,
> 
> TBH I don't really think it's a big deal, since `docs` is also very big, and we didn't make it `taichi-docs`. Also note that the size of `baseline` depends only on `tests`.
> In contrast, not having submodule for `baseline` have pros, like the `docs` does - we can update `docs` in each PR belong to that.

Your concerns are very valid. Here are my thoughts:

It's true that 100+KB is not too big, but I would worry that in the future we might need to update these (machine-generated) files very frequently. Also reviewing a changeset with 370 files is not easy... Note that almost every file will be changed if we change IR optimizations. 

It's also true that docs are big, but humans write them and every time the changeset won't be too big.



> Note that with this attempt, the PR author can't update `baseline` themselves since they may don't have access to `taichi-baseline`.

I feel like performance-related PRs will mostly be made by people who are familiar enough with the code base, so it's reasonable to grant them access to `taichi-baseline`. WDTY?



> Again, I dislike the submodule attempt, although it helps disable conflict when benchmark dats get merged.

I don't like submodules either, but unfortunately, the design here is hard to make perfect I guess...

**Comment by archibate** [comment]
> It's also true that docs are big, but humans write them and every time the changeset won't be too big.

I got the point, thank you!

> I feel like performance-related PRs will mostly be made by people who are familiar enough with the code base, so it's reasonable to grant them access to taichi-baseline. WDTY?

Sounds reasonable, let's merge this first.

**Pull request merged**